gnomad_qc/v4/annotations/generate_freq.py

```diff
@@ -0,0 +1,1027 @@
+        **hl.if_else(
+            ht.freq[0].AF > af_threshold,
+            hl.struct(
+                ab_adjusted_freq=call_stats_expr,
+                **qual_hist_expr,
+                ab_adjusted_age_hists=ht.high_ab_het_adjusted_age_hists,
+            ),
+            hl.struct(**no_ab_adjusted_expr),
+        ),
+    )
+
+    return ht
+
+
+def generate_faf_grpmax(ht: hl.Table) -> hl.Table:
+    """
+    Compute filtering allele frequencies ('faf') and 'grpmax' with the AB-adjusted frequencies.
+
+    :param ht: Hail Table containing 'freq', 'ab_adjusted_freq', 'high_ab_het'
+        annotations.
+    :return: Hail Table with 'faf' and 'grpmax' annotations.
+    """
+    logger.info(
+        "Filtering frequencies to just 'non_ukb' subset entries for 'faf' "
+        "calculations..."
+    )
+    # Remove the key "subset" from each dict in 'non_ukb_freq_meta' list so 'faf' will
+    # pull the correct indices, i.e. `faf_expr` requires specific lengths to consider
+    # each element in the freq list.
+    non_ukb_ht = filter_freq_arrays_for_non_ukb_subset(
+        ht,
+        items_to_filter={"subset": ["non_ukb"]},
+        combine_operator="or",
+        annotations=("ab_adjusted_freq",),
+        remove_subset_from_meta=True,
+    )
+    freq_metas = [
+        (ht.ab_adjusted_freq, ht.index_globals().freq_meta),
+        (non_ukb_ht[ht.key].ab_adjusted_freq, non_ukb_ht.index_globals().freq_meta),
+    ]
+    faf_grpmax_expr = [
+        hl.struct(
+            **hl.bind(
+                lambda f, g: hl.struct(
+                    faf=f[0],
+                    faf_meta=f[1],
+                    grpmax=g.annotate(
+                        faf95=f[0][
+                            f[1].index(lambda y: y.values() == ["adj", g.pop])
+                        ].faf95
+                    ),
+                ),
+                faf_expr(freq, meta, ht.locus, POPS_TO_REMOVE_FOR_POPMAX),
+                pop_max_expr(freq, meta, POPS_TO_REMOVE_FOR_POPMAX),
+            )
+        )
+        for freq, meta in freq_metas
+    ]
+
+    logger.info("Annotating 'faf' and 'grpmax'...")
+    ht = ht.annotate(
+        faf=[group for dataset in faf_grpmax_expr for group in dataset],
+        grpmax=[dataset.grpmax for dataset in faf_grpmax_expr],
+    )
+    faf_meta_expr = [x.faf_meta.collect(_localize=False)[0] for x in faf_grpmax_expr]
+    # Add subset back to non_ukb faf meta and flatten it
+    faf_meta_expr[1] = faf_meta_expr[1].map(
+        lambda d: hl.dict(d.items().append(("subset", "non_ukb")))
+    )
+    ht = ht.annotate_globals(
+        faf_meta=hl.flatten(faf_meta_expr),
+        faf_index_dict=[
+            make_faf_index_dict(hl.eval(x), label_delimiter="-") for x in faf_meta_expr
+        ],
+        grpmax_meta=[
+            {"dataset": "gnomad"},
+            {"dataset": "non_ukb"},
+        ],  # TODO: These seem silly but keeps with the meta/dict theme of globals
+        grpmax_index_dict=SUBSET_DICT,
```

Comment:
Sorry, I commited a suggestion on name and it looks like it removed @jkgoodrich 's other comment. @ch-kr do you have thoughts on the grpmax_meta and index_dict? L778-782

Reply:
hmmm. what about removing these globals and restructuring grpmax to have the same structure as the age hists? something like
```
grpmax: struct{
    gnomad: struct{
        AC: int32,
        AF: float64,
        AN: int32,
        homozygote_count: int64,
        pop: str,
        faf95: float64,
    },
    non-ukb: struct{
            AC: int32,
            AF: float64,
            AN: int32,
            homozygote_count: int64,
            pop: str,
            faf95: float64,
        }
} 
```

Reply:
I'm fine with that but just confirming we're dropping age hists so it actually wont look like the above?

Reply:
oh yes sorry that wasn't clear. we're dropping the non-UKB age hists, so that annotation looks like this now, right?
```
age_hists: struct{
            age_hist_hom: struct {
                bin_edges: array<float64>, 
                bin_freq: array<int64>, 
                n_smaller: int64, 
                n_larger: int64
            }, 
            age_hist_het: struct {
                bin_edges: array<float64>, 
                bin_freq: array<int64>, 
                n_smaller: int64, 
                n_larger: int64
            },
        }
```